addons/controllers/cpi/cpiconfig_utils.go

```diff
@@ -0,0 +1,307 @@
+	pkgtypes "github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	cpiv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/cpi/v1alpha1"
+)
+
+// mapCPIConfigToDataValuesNonParavirtual generates CPI data values for non-paravirtual modes
+func (r *CPIConfigReconciler) mapCPIConfigToDataValuesNonParavirtual( // nolint
+	ctx context.Context,
+	cpiConfig *cpiv1alpha1.CPIConfig, cluster *clusterapiv1beta1.Cluster) (*CPIDataValues, error,
+) { // nolint:whitespace
+	d := &CPIDataValues{}
+	c := cpiConfig.Spec.VSphereCPI.NonParavirtualConfig
+	d.VSphereCPI.Mode = CPINonParavirtualMode
+
+	// get the vsphere cluster object
+	vsphereCluster, err := r.getVSphereCluster(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	// derive the thumbprint, server from the vsphere cluster object
+	d.VSphereCPI.TLSThumbprint = vsphereCluster.Spec.Thumbprint
+	d.VSphereCPI.Server = vsphereCluster.Spec.Server
+
+	// derive vSphere username and password from the <cluster name> secret
+	clusterSecret, err := getSecret(ctx, r.Client, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return nil, err
+	}
+	d.VSphereCPI.Username, d.VSphereCPI.Password, err = getUsernameAndPasswordFromSecret(ctx, r.Client, clusterSecret)
+	if err != nil {
+		return nil, err
+	}
+
+	// get the control plane machine template
+	cpMachineTemplate := &capvv1beta1.VSphereMachineTemplate{}
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      controlPlaneName(cluster.Name),
+	},
+		cpMachineTemplate); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.Errorf("VSphereMachineTemplate %s/%s not found", cluster.Namespace, controlPlaneName(cluster.Name))
+		}
+		return nil, errors.Errorf("VSphereMachineTemplate %s/%s could not be fetched, error %v", cluster.Namespace, controlPlaneName(cluster.Name), err)
+	}
+
+	// derive data center information from control plane machine template, if not provided
+	d.VSphereCPI.Datacenter = cpMachineTemplate.Spec.Template.Spec.Datacenter
+
+	// derive ClusterCidr from cluster.spec.clusterNetwork
+	if cluster.Spec.ClusterNetwork != nil && cluster.Spec.ClusterNetwork.Pods != nil && len(cluster.Spec.ClusterNetwork.Pods.CIDRBlocks) > 0 {
+		d.VSphereCPI.Nsxt.Routes.ClusterCidr = cluster.Spec.ClusterNetwork.Pods.CIDRBlocks[0]
+	}
+
+	// derive IP family or proxy related settings from cluster annotations
+	if cluster.Annotations != nil {
+		d.VSphereCPI.IPFamily = cluster.Annotations[pkgtypes.IPFamilyConfigAnnotation]
+		d.VSphereCPI.HTTPProxy = cluster.Annotations[pkgtypes.HTTPProxyConfigAnnotation]
+		d.VSphereCPI.HTTPSProxy = cluster.Annotations[pkgtypes.HTTPSProxyConfigAnnotation]
+		d.VSphereCPI.NoProxy = cluster.Annotations[pkgtypes.NoProxyConfigAnnotation]
+	}
+
+	// derive nsxt related configs from cluster variable
+	d.VSphereCPI.Nsxt.PodRoutingEnabled = r.tryParseClusterVariableBool(ctx, cluster, "NSXT_POD_ROUTING_ENABLED")
```

Comment:
Can we define these cluster variable names as constants and store somewhere?

Reply:
I will put them in `cpiconfig_const.go`